app/Actions/Diagnostics/Checks/ImageOptCheck.php

```diff
@@ -36,9 +37,11 @@ public function check(array &$errors): void
 		}
 
 		foreach ($tools as $tool) {
-			$path = exec('command -v ' . $binaryPath . $tool->binaryName());
-			if ($path === '') {
-				$errors[] = 'Warning: lossless_optimization set to 1 but ' . $binaryPath . $tool->binaryName() . ' not found!';
+			if (Helpers::isExecAvailable()) {
+				$path = exec('command -v ' . $binaryPath . $tool->binaryName());
+				if ($path === '') {
+					$errors[] = 'Warning: lossless_optimization set to 1 but ' . $binaryPath . $tool->binaryName() . ' not found!';
+				}
 			}
```

Comment:
I wonder if should actually check for a disabled `exec` in each iteration again and log one message for each iteration or if we should pull out the test for `exec`. In particular the way it is now, the user cannot distinguish between the case that the tool itself is missing and the case that the tool is yet there but `exec` is disabled. Users might get confused by this, if the error message (falsely) claims that the tool is missing.

Reply:
> Users might get confused by this, if the error message (falsely) claims that the tool is missing.

But the error is only raised if exec is there? I haven't checked - does no `exec` raise a warning at the top of the page?

> or if we should pull out the test for `exec`.

That makes sense to me, putting the `if (Helpers::isExecAvailable()) {` outside the `foreach`.

Reply:
Ah, sorry my fault. You are right, the error message is only logged _if_ `exec` is available. My fault. I reviewed this on my smartphone and the Github Android app is really bad at presenting diffs properly. I misread the curly braces.

Yes, the check for exec should be outside the loop.